libp2p/stream/bufferstream.nim

```diff
@@ -83,7 +75,23 @@ proc setupBufferStreamTracker(): BufferStreamTracker =
   result.dump = dumpTracking
   result.isLeaked = leakTransport
   addTracker(BufferStreamTrackerName, result)
-declareGauge libp2p_open_bufferstream, "open BufferStream instances"
+
+type
+  # TODO: figure out how to make this generic to avoid casts
+  WriteHandler* = proc (data: seq[byte]): Future[void] {.gcsafe.}
+
+  BufferStream* = ref object of LPStream
+    maxSize*: int                     # buffer's max size in bytes
+    readBuf: Deque[byte]              # this is a ring buffer based dequeue
+    readReqs*: Deque[Future[void]]    # use dequeue to fire reads in order
+    dataReadEvent*: AsyncEvent        # event triggered when data has been consumed from the internal buffer
+    writeHandler*: WriteHandler       # user provided write callback
+    writeLock*: AsyncLock             # write lock to guarantee ordered writes
+    lock: AsyncLock                   # pushTo lock to guarantee ordered reads
+    piped: BufferStream               # a piped bufferstream instance
```

Comment:
the pipe stuff, is it actively used?

Reply:
it's used in tests, can probably be moved to some utils.